platform/android/llapp_main.lua

```diff
@@ -15,6 +15,72 @@ end
 -- path to primary external storage partition
 local path = android.getExternalStoragePath()
 
+-- execute user scripts before patch.lua
+local lfs = require("libs/libkoreader-lfs")
+
+local user_dir = android.getExternalStoragePath() .. "/koreader"
+local afterupdate_marker = android.dir .. "/afterupdate.marker"
+local run_once_scripts = user_dir .. "/scripts.afterupdate"
+local run_always_scripts = user_dir .. "/scripts.always"
+
+local shell_scripts = {}
+
+local function getUserScripts(path)
+    shell_scripts = {} -- zero list, as getUserScripts could be called multiple times
+    local ret = lfs.attributes(path)
+    if ret == nil then
+        android.LOGI("no script folder " .. path .. " found!")
+        return
+    end
+    if ret.mode == "directory" then
+        android.LOGI("script folder " .. path .. " found!")
+        for entry in lfs.dir(path) do
+            if entry ~= "." and entry ~= ".." then
+                local fullpath = path .. "/" .. entry
+                if lfs.attributes(fullpath).mode ~= "directory" then
+                    if fullpath:match(".sh$") then  -- only include files ending in ".sh"
+                        table.insert(shell_scripts, fullpath)
+                    end
+                else
+                    getUserScripts(fullpath) -- recurse into next directory
+                end
+            end
+        end
+    end
+end
+
+local function runUserScripts(scripts)
+    for _, script in ipairs(scripts) do
+        local ret = os.execute("/system/bin/sh " .. script, path, android.dir)
+        if ret == 0 then
+            android.LOGI("script " .. script  .. " executed succesfully")
+        else
+            android.LOGW("failed to execute " .. script)
+        end
+    end
+end
+
+-- scripts executed once after an update of koreader
+android.LOGI("checking and running scripts on update, if necessary.")
+if lfs.attributes(run_once_scripts, "mode") == "directory" then
+    if lfs.attributes(afterupdate_marker, "mode") == "file" then
+        getUserScripts(run_once_scripts)
+        runUserScripts(shell_scripts)
+        android.LOGI(string.format("Executed %d afterupdate scripts from %s", #shell_scripts, run_once_scripts))
+        shell_scripts = {}
+        android.execute("rm", afterupdate_marker)
```

Comment:
Most android devices don't have a sane PATH. Use the fullpath to the remove command `/system/bin/rm`

Reply:
Good to know. Then we have to change the two ```chmod``` a few lines further down, too.

Reply:
Yup, I think you got a point here. chmod never fails so rm shouldn't fail. Sorry :p